### PR TITLE
feat: implement recursive download for dfget

### PR DESCRIPTION
**Pull request by Lzzzzzt**
implement recursive download for dfget

## Description

This PR completes with the following tasks:

- [x] add new function called `download_tasks` to download the directory
- [x] add new function called `download_task` to download single file

Other details will be shown in the code. 

## Related Issue

#496 

## Motivation and Context

Dragonfly rust client does not support back-to-source downloading of different object storage protocols. Priority is given to supporting OSS protocol back-to-source downloads. Need to support OSS protocol in dragonfly-client-backend crate, implemented based on trait Backend and it needs to support directory recursive downloading.

## Screenshots

image

### CODE REVIEW

File: dragonfly-client/src/bin/dfget/main.rs [hunk 1]
```diff
@@ -448,37 +466,182 @@ async fn main() -> anyhow::Result<()> {
...
+
+    // Use the semaphore to control the concurrent download task number.
+    // The initial value of semaphore is taken from the user input.
+    let concurrent_control = Arc::new(Semaphore::new(args.download_concurrent_count));
+
+    // To store the download task handler.
+    let mut handlers = Vec::with_capacity(file_count);
+
+    for entry in entries {
+        let url: Url = entry.url.parse().expect("unexpected url");
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:575]: How about returning an error instead of panic?

File: dragonfly-client/src/bin/dfget/main.rs [hunk 2]
```diff
@@ -195,6 +199,20 @@ struct Args {
     )]
     storage_predefined_acl: Option<String>,
 
+    #[arg(
+        long,
+        default_value_t = 8,
+        help = "Specify the max number of file to download"
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:205]: Specify the `maximum` number of `files` to download.

File: dragonfly-client/src/bin/dfget/main.rs [hunk 3]
```diff
@@ -195,6 +199,20 @@ struct Args {
...
+    )]
+    download_max_files: usize,
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:207]: Unusual naming.
image

File: dragonfly-client/src/bin/dfget/main.rs [hunk 4]
```diff
@@ -195,6 +199,20 @@ struct Args {
...
+
+    #[arg(
+        long,
+        default_value_t = 5,
+        help = "Specify the concurrent count of download task"
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:212]: tasks

File: dragonfly-client/src/bin/dfget/main.rs [hunk 5]
```diff
@@ -448,37 +466,182 @@ async fn main() -> anyhow::Result<()> {
...
             err
         })?;
 
+    // Get the absolute path of the output file.
+    let absolute_path = Path::new(&args.output).absolutize()?;
+    info!("download file to: {}", absolute_path.to_string_lossy());
+
+    args.output = absolute_path.into();
+
+    // If url end with '/', treat it as a directory and download the whole directory.
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:483]: ends

File: dragonfly-client/src/bin/dfget/main.rs [hunk 6]
```diff
@@ -448,37 +466,182 @@ async fn main() -> anyhow::Result<()> {
...
+            endpoint: args.storage_endpoint.clone(),
+            credential: args.storage_credential.clone(),
+            predefined_acl: args.storage_predefined_acl.clone(),
         });
     }
 
-    // Get the absolute path of the output file.
-    let absolute_path = Path::new(&args.output).absolutize()?;
-    info!("download file to: {}", absolute_path.to_string_lossy());
+    // Init the backend factory to choose which backend to use for send head request.
```
Comment (BruceAko) [inline dragonfly-client/src/bin/dfget/main.rs:512]: sending

**Review by BruceAko — CHANGES_REQUESTED** [review]

**Comment by BruceAko** [comment]
Redundant signature information.
image

File: dragonfly-client/src/bin/dfget/main.rs (hunk 3 above)
Comment (Lzzzzzt) [reply dragonfly-client/src/bin/dfget/main.rs:207]: this follow the other naming in the dfget

File: dragonfly-client/src/bin/dfget/main.rs (hunk 2 above)
Comment (Lzzzzzt) [reply dragonfly-client/src/bin/dfget/main.rs:205]: follow by the other `help` in the dfget

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**